agent.go

```diff
@@ -200,6 +206,99 @@ func (c *container) getProcess(execID string) (*process, error) {
 	return proc, nil
 }
 
+// setSandboxStorage sets the sandbox level reference
+// counter for the sandbox storage.
+// This method also returns a boolean to let
+// callers know if the storage already existed or not.
+// It will return true if storage is new.
+//
+// It's assumed that caller is calling this method after
+// acquiring a lock on sandbox.
+func (s *sandbox) setSandboxStorage(path string) bool {
+	if _, ok := s.storages[path]; !ok {
+		sbs := &sandboxStorage{refCount: 1}
+		s.storages[path] = sbs
+		return true
+	}
+	sbs := s.storages[path]
+	sbs.Lock()
+	sbs.refCount++
+	sbs.Unlock()
+	return false
+}
+
+// unSetSandboxStorage will decrement the sandbox storage
+// reference counter. If there aren't any containers using
+// that sandbox storage, this method will also remove the
+// storage from the sandbox.
+//
+// It's assumed that caller is calling this method after
+// acquiring a lock on sandbox.
+func (s *sandbox) unSetSandboxStorage(path string) error {
+	if sbs, ok := s.storages[path]; ok {
+		sbs.Lock()
+		defer sbs.Unlock()
+
+		sbs.refCount--
+
+		// If this sandbox storage is not used by any container
+		// then remove it's reference
+		if sbs.refCount < 1 {
+			delete(s.storages, path)
+		}
+		return nil
+	}
+	return grpcStatus.Errorf(codes.NotFound, "Sandbox storage with path %s not found", path)
+}
+
+// removeSandboxStorage removes the sandbox storage if no
+// containers are using that storage.
+//
+// It's assumed that caller is calling this method after
+// acquiring a lock on sandbox.
+func (s *sandbox) removeSandboxStorage(path string) error {
+	err := removeMounts([]string{path})
+	if err != nil {
+		return grpcStatus.Errorf(codes.Unknown, "Unable to unmount sandbox storage path %s", path)
+	}
+	err = os.RemoveAll(path)
+	if err != nil {
+		return grpcStatus.Errorf(codes.Unknown, "Unable to delete sandbox storage path %s", path)
+	}
+	return nil
+}
+
+// unsetAndRemoveSandboxStorage unsets the storage from sandbox
+// and if there are no containers using this storage it will
+// remove it from the sandbox
+//
+// It's assumed that caller is calling this method after
+// acquiring a lock on sandbox.
+func (s *sandbox) unsetAndRemoveSandboxStorage(path string) error {
+	if sbs, ok := s.storages[path]; ok {
+		if err := s.unSetSandboxStorage(path); err != nil {
```

Comment:
`unSetSandboxStorage` already removes the sbs from the map. So I'd prefer it to return a boolean to indicate that the sbs needs to be cleaned up. Then you do not need to check the refcount again below.

Reply:
Yes that's a good point, the boolean will save some code, but I would like to keep the error return too, so I think the best thing to do is:
```go
func (s *sandbox) unSetSandboxStorage(path string) (bool, error) {
```